### PR TITLE
Convert Private Orgs to `userRegistrationRequired` Orgs

### LINKED ISSUE #1604: API - Remove Support for Private Organizations
**Is your feature request related to a problem? Please describe.**
We need to reduce and simplify some of the Talawa features. 

The concept of private organizations is unlikely given that most organizations using Talawa will be non-profits where transparency will be important.

**Describe the solution you'd like**

Remove all support for private organizations in:
1. Backend systems and supporting code

Other:

1. All tests must be valid and pass
1. No other functionality must be affected.

**Describe alternatives you've considered**
- N/A

**Approach to be followed (optional)**
- See above

**Additional context**
Related issues include:
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1604
2. https://github.com/PalisadoesFoundation/talawa-admin/issues/1327
3. https://github.com/PalisadoesFoundation/talawa/issues/2287

**Potential internship candidates**
- Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by AVtheking**
What kind of change does this PR introduce?
Removes support for private organizations

Issue Number:

Fixes #1604 

Did you add tests for your changes?
NO

Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?
Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
@AVtheking 

When Talawa was first created it was thought that many unrelated organizations would share the same API instance. Each organization would allow anyone to register with them automatically, just like you can with most online services.

We also thought that some organizations would want to screen users before accepting registrations. These could be private clubs with membership requirements. This was the inspiration for the `isPrivate` flag in the API schema.

We then thought that some of these clubs may not want to be visible in the list of organizations under certain circumstances. This is why we have the `visibleInSearch` flag.

We have decided to assume that each Talawa instance will manage related organizations and that new features will focus on non-profits where openness is cherished. The concept of a `isPrivate` didn’t make sense in that context.

We are trying to clean up the code in 2024 while simultaneously adding many new features, and we decided to remove private organizations.

However, as many of you have rightly pointed out, there will still be a reasonable chance that Admins will not want to have open enrollment with their organizations and will want to do some vetting of new  user applicants especially if it is done outside of the event checkin process.

We could use the `isPrivate` flag for this registration/enrollment required toggle, but the name will cause confusion for future contributors and end users.

I propose that we rename the flag `userRegistrationRequired` which is unambiguous and retain the `visibleInSearch` flag until further notice.

We have a registration approval screen redesign that will benefit from this clarification too.

Wherever we reference private organizations in the UI, the wording should reflect this new approach, asking whether new users need to register instead of whether the organization is private.

For Talawa-Admin, the default must be for registration to be `required` as it is the more conservative security approach.

This product direction will hopefully clarify the uncertainty.

**Comment by AVtheking** [comment]
@palisadoes  I have changed the isPublic flag to userRegisterationRequired flag as you have suggested

**Comment by palisadoes** [comment]
Thanks. I've assigned reviewers. Please work with them to get the code checked and merged

**Comment by AVtheking** [comment]
ok sir

**Comment by xoldyckk** [comment]
if not arguing semantics `userRegistrationRequired: boolean` or `isPublic: true` or  `isPrivate: true` etc would all mean the same thing, dunno see how this would confuse contributors

File: .husky/pre-commit [hunk 1]
```diff
@@ -18,8 +18,12 @@ npm run lint:fix
...
 npm run gen:schema
 
-# Generate table of contents for markdown files in root directory
-npm run update:toc
+# Check if any markdown files have been changed
+if git diff --cached --name-only | grep -E '\.md$'
+then
+  # Generate table of contents for markdown files in root directory
+  npm run update:toc
+fi
```
Comment (xoldyckk) [inline .husky/pre-commit:27]: don't add unrelated changes like these to the PR, because there's no context on what's being changed here and why

File: package.json [hunk 2]
```diff
@@ -105,9 +104,9 @@
     "@types/cors": "^2.8.17",
     "@types/express": "^4.17.17",
     "@types/express-rate-limit": "^5.1.3",
-    "@types/graphql-depth-limit": "^1.1.6",
-    "@types/i18n": "^0.13.10",
-    "@types/jsonwebtoken": "^9.0.5",
+    "@types/graphql-depth-limit": "^1.1.3",
+    "@types/i18n": "^0.13.3",
```
Comment (xoldyckk) [inline package.json:108]: why are these packages being downgraded? sync your local repo with develop branch of talawa-api main repo and make a new commit

File: src/models/Organization.ts [hunk 3]
```diff
@@ -25,9 +24,10 @@ export interface InterfaceOrganization {
   pinnedPosts: PopulatedDoc<InterfacePost & Document>[];
   membershipRequests: PopulatedDoc<InterfaceMembershipRequest & Document>[];
   blockedUsers: PopulatedDoc<InterfaceUser & Document>[];
-  visibleInSearch: boolean | undefined;
   customFields: PopulatedDoc<InterfaceOrganizationCustomField & Document>[];
   createdAt: Date;
+  userRegistrationRequired: boolean;
+  visibleInSearch: boolean | undefined;
```
Comment (xoldyckk) [inline src/models/Organization.ts:30]: remove the undefined

File: src/models/Organization.ts [hunk 4]
```diff
@@ -66,10 +65,13 @@ const organizationSchema = new Schema({
   location: {
     type: String,
   },
-  isPublic: {
+  userRegistrationRequired: {
     type: Boolean,
     required: true,
   },
+  visibleInSearch: {
```
Comment (xoldyckk) [inline src/models/Organization.ts:72]: set the default value of `visibleInSearch` to true

File: src/typeDefs/inputs.ts [hunk 5]
```diff
@@ -290,9 +288,9 @@ export const inputs = gql`
   input UpdateOrganizationInput {
     name: String
     description: String
-    isPublic: Boolean
-    visibleInSearch: Boolean
     location: String
+    userRegistrationRequired: Boolean!
```
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:292]: make the newly added fields nullable in `UpdateOrganizationInput`

**Review by xoldyckk — COMMENTED** [review]
fix the tests

**Comment by AVtheking** [comment]
I have made the changes that are requested by you

**Comment by palisadoes** [comment]
What changes? There is no history

**Comment by AVtheking** [comment]
changes which are suggested by @xoldyckk to me yesterday

**Comment by palisadoes** [comment]
> changes which are suggested by @xoldyckk to me yesterday

Where? There is no history in this PR

File: src/typeDefs/inputs.ts [hunk 6]
```diff
@@ -146,10 +146,10 @@ export const inputs = gql`
     description: String!
     location: String
     attendees: String
-    isPublic: Boolean!
-    visibleInSearch: Boolean!
     apiUrl: URL
     image: String
+    userRegistrationRequired: Boolean!
+    visibleInSearch: Boolean!
```
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:152]: both these fields don't need to be required in the input as database by default will assign them a sane default value, they're opt in fields

File: src/models/Organization.ts [hunk 7]
```diff
@@ -66,10 +65,14 @@ const organizationSchema = new Schema({
   location: {
     type: String,
   },
-  isPublic: {
+  userRegistrationRequired: {
     type: Boolean,
     required: true,
```
Comment (xoldyckk) [inline src/models/Organization.ts:70]: assign this a default value of `false`

File: src/models/Organization.ts [hunk 8]
```diff
@@ -66,10 +65,14 @@ const organizationSchema = new Schema({
...
   },
+  visibleInSearch: {
+    type: Boolean,
+    default: true,
+  },
```
Comment (xoldyckk) [inline src/models/Organization.ts:75]: `required: true`

File: src/resolvers/Mutation/joinPublicOrganization.ts [hunk 9]
```diff
@@ -49,15 +48,6 @@ export const joinPublicOrganization: MutationResolvers["joinPublicOrganization"]
       );
     }
 
-    // Checks whether organization is public.
-    if (organization.isPublic === false) {
```
Comment (xoldyckk) [inline src/resolvers/Mutation/joinPublicOrganization.ts:53]: shouldn't this check be replaced with `organization.userRegistrationRequired === true` instead of being removed?

**Review by xoldyckk — COMMENTED** [review]
@AVtheking look up `git squash`, your commit history is too cluttered and unorganized, when you're making small insignificant changes and committing rapidly you can squash those commits into one meaningful commit

**Comment by AVtheking** [comment]
@xoldyckk  I have made the changes requested by you.

File: tests/resolvers/Mutation/joinPublicOrganization.spec.ts [hunk 10]
```diff
@@ -72,70 +71,6 @@ describe("resolvers -> Mutation -> joinPublicOrganization", () => {
     }
   });
```
Comment (xoldyckk) [inline tests/resolvers/Mutation/joinPublicOrganization.spec.ts:73]: you forgot the test cases

File: src/resolvers/Mutation/joinPublicOrganization.ts [hunk 11]
```diff
@@ -49,19 +49,18 @@ export const joinPublicOrganization: MutationResolvers["joinPublicOrganization"]
...
+    const currentUserExists = await User.exists({
+      _id: context.userId,
+    });
+
+    if (organization.userRegistrationRequired == false) {
```
[the remark this answers is not in the document]
Comment (AVtheking) [reply src/resolvers/Mutation/joinPublicOrganization.ts]: ok sir ,thanks for the advice

File: tests/resolvers/Mutation/joinPublicOrganization.spec.ts (hunk 10 above)
Comment (AVtheking) [reply tests/resolvers/Mutation/joinPublicOrganization.spec.ts]: added the test case

**Comment by palisadoes** [comment]
@AVtheking Is this ready for another round of reviews?

**Comment by AVtheking** [comment]
Yes sir

**Comment by palisadoes** [comment]
- @xoldyckk
- @sumitra19jha
- @ttaylor92
- @SiddheshKukade
- @rishav-jha-mech
- @noman2002

Please review once more. It's ready

**Review by xoldyckk — APPROVED** [review]

**Review by palisadoes — APPROVED** [review]
